<launch>
    <arg name="uuv_name" default="eca_a9"/>
    <arg name="model_name" default="$(arg uuv_name)"/>
    <arg name="gui_on" default="true"/>
    <arg name="use_params_file" default="false"/>

    <arg name="use_ned_frame" default="false"/>
<!-- 
    <arg name="teleop_on" default="false"/>
    <arg name="joy_id" default="0"/>
    <arg name="axis_yaw" default="0"/>
    <arg name="axis_x" default="4"/>
    <arg name="axis_y" default="3"/>
    <arg name="axis_z" default="1"/> -->

    <arg name="Kd" default="24.328999405818507,95.16574836816616,25.943377407248825,0,6.388371356010936,79.2844976871164"/>
    <arg name="Ki" default="0.0010232768152540483,0.0010232768152540483,0.0010232768152540483,0,0.11901644069756079,0.11901644069756079"/>
    <arg name="slope" default="0.9903858668992097,0.9903858668992097,0.9903858668992097,0,0.20796465986893387,0.20796465986893387"/>

   <!--  
    
    
    <arg name="n_fins" default="4"/>
    <arg name="map_roll" value="1,1,1,1"/>
    <arg name="map_pitch" value="1,1,-1,-1"/>
    <arg name="map_yaw" value="-1,1,1,-1"/>   
    <arg name="max_fin_angle" default="1.396263402"/>
    <arg name="idle_radius" default="30"/> -->  

    <arg name="look_ahead_delay" default="0.0"/>
    <arg name="min_thrust" default="0"/>
    <arg name="max_thrust" default="120"/> 

    <include file="$(find eca_a9_control)/launch/start_control_allocator.launch"/>
    <include file="$(find eca_a9_control)/launch/start_eca_a9_performance_logger.launch"/> 

    <!-- Thruster configuration -->
    <!-- <arg name="thruster_topic" default="thrusters/0/input"/>
    <arg name="thruster_topic_prefix" default="thrusters" />
    <arg name="thruster_topic_suffix" default="input" />
    <arg name="thruster_frame_base" default="thruster_" />
    <arg name="thruster_conversion_fcn" default="proportional" /> -->
    <!-- For proportional thruster model: gain * rotor_ang_vel * |rotor_ang_vel| -->
    <!-- <arg name="thruster_gain" default="0.000049"/>

    <arg name="fin_topic_prefix" default="fins"/>
    <arg name="fin_topic_suffix" default="input"/>  
 -->
    <group if="$(arg gui_on)">
      <group ns="$(arg uuv_name)">
          <node pkg="uuv_control_utils"
              type="trajectory_marker_publisher.py"
              name="trajectory_marker_publisher"
              output="screen"/>
              <remap from="trajectory" to="dp_controller/trajectory"/>
              <remap from="waypoints" to="dp_controller/waypoints"/>
              <remap from="reference" to="dp_controller/reference"/>
      </group>
    </group>

    <group ns="$(arg uuv_name)">
      <node pkg="uuv_trajectory_control"
            type="rov_nmb_sm_controller.py"
            name="rov_nmb_sm_controller"
            output="screen"
            if="$(arg use_ned_frame)">
        <!-- Remap necessary topics -->
        <remap from="odom" to="pose_gt_ned"/>
        <remap from="trajectory" to="dp_controller/trajectory"/>
        <remap from="input_trajectory" to="dp_controller/input_trajectory"/>
        <remap from="waypoints" to="dp_controller/waypoints"/>
        <remap from="error" to="dp_controller/error"/>
        <remap from="reference" to="dp_controller/reference"/>
        <remap from="thruster_output" to="thruster_manager/input_stamped"/>
        <remap from="auv_command_output" to="control_allocation/control_input"/>
        <!-- Controller parameters -->
        <rosparam subst_value="true">
          saturation: 200
          thrusters_only: false
          Kd: [$(arg Kd)]
          Ki: [$(arg Ki)]
          slope: [$(arg slope)]
          inertial_frame_id: world_ned
          max_forward_speed: 2.0
          use_stamped_poses_only: false
          look_ahead_delay: $(arg look_ahead_delay)
          min_thrust: $(arg min_thrust)
          max_thrust: $(arg max_thrust)
          is_underactuated: true
          dubins: 
            radius: 10
            max_pitch: 0.26
            desired_pitch_limit: 0.26
        </rosparam>
      </node>

      <node pkg="uuv_trajectory_control"
            type="rov_nmb_sm_controller.py"
            name="rov_nmb_sm_controller"
            output="screen"
            unless="$(arg use_ned_frame)">
        <!-- Remap necessary topics -->
        <remap from="odom" to="pose_gt"/>
        <remap from="trajectory" to="dp_controller/trajectory"/>
        <remap from="input_trajectory" to="dp_controller/input_trajectory"/>
        <remap from="waypoints" to="dp_controller/waypoints"/>
        <remap from="error" to="dp_controller/error"/>
        <remap from="reference" to="dp_controller/reference"/>
        <remap from="thruster_output" to="thruster_manager/input_stamped"/>
        <remap from="auv_command_output" to="control_allocation/control_input"/>
        <!-- Controller parameters -->
        <rosparam subst_value="true">
          saturation: 200
          thrusters_only: false
          Kd: [$(arg Kd)]
          Ki: [$(arg Ki)]
          slope: [$(arg slope)]
          inertial_frame_id: world
          max_forward_speed: 2.0
          use_stamped_poses_only: false
          look_ahead_delay: $(arg look_ahead_delay)
          min_thrust: $(arg min_thrust)
          max_thrust: $(arg max_thrust)
          is_underactuated: true
          dubins: 
            radius: 10
            max_pitch: 0.26
            desired_pitch_limit: 0.26
        </rosparam>
      </node>
    </group>

    <group if="$(arg gui_on)">
        <node name="rviz" pkg="rviz" type="rviz" output="screen" args="-d $(find eca_a9_control)/rviz/eca_a9_control.rviz"/>
        <!-- <remap from="sss_right" to="eca_a9/sonar_right"/> -->
    </group>
</launch>
